ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+from abc import abstractmethod
+import collections
+import io
+import json
+import logging
+import pathlib
+from typing import Any, Dict, Generic, Mapping, Optional, Tuple, TypeVar, cast
+
+from async_service import Service
+from eth_utils import ValidationError
+import trio
+
+from ddht.abc import RPCHandlerAPI, RPCRequest, RPCResponse
+from ddht.typing import JSON
+
+NEW_LINE = "\n"
+
+
+def strip_non_json_prefix(raw_request: str) -> Tuple[str, str]:
+    if raw_request and raw_request[0] != "{":
+        prefix, bracket, rest = raw_request.partition("{")
+        return prefix.strip(), bracket + rest
+    else:
+        return "", raw_request
+
+
+logger = logging.getLogger("ddht.rpc")
+decoder = json.JSONDecoder()
+
+
+async def read_json(
+    socket: trio.socket.SocketType,
+    buffer: io.StringIO,
+    decoder: json.JSONDecoder = decoder,
+) -> JSON:
+    request: JSON
+
+    while True:
+        data = await socket.recv(1024)
+        buffer.write(data.decode())
+
+        bad_prefix, raw_request = strip_non_json_prefix(buffer.getvalue())
+        if bad_prefix:
+            logger.info("Client started request with non json data: %r", bad_prefix)
+            await write_error(socket, f"Cannot parse json: {bad_prefix}")
+            continue
```

Comment:
I'm not sure in which situations the client will give us non-JSON data, but it seems unlikely that we'll receive any more requests which make sense, this might want to kill the connection!

And if it doesn't kill the connection, I think it should clear the buffer, right now the buffer is left as-is, so we'll loop and throw the same failure every time we receive more data.